File: backend/src/services/pipeline.ts
/**
 * Shared gather -> extract -> synthesize pipeline used by all four modules.
 *
 * 1. gather: call the web-data source (through cachedFetch, so repeat demo
 *    runs don't burn credits) to get raw web data.
 * 2. extract: cheap AIML model turns raw data into structured facts.
 * 3. synthesize: reasoning AIML model turns structured facts + profile context
 *    into the module's final output.
 */

import { createHash } from "node:crypto";

import { eq } from "drizzle-orm";

import { db } from "../db/index.js";
import { scrapeCache } from "../db/schema.js";
import * as aimlClient from "./aimlClient.js";

function cacheKey(toolName: string, query: string): string {
  return createHash("sha256").update(`${toolName}:${query}`).digest("hex");
}

/** Look up (toolName, query) in scrape_cache; call fetchFn and store on miss. */
export async function cachedFetch<T>({
  toolName,
  query,
  fetchFn,
}: {
  toolName: string;
  query: string;
  fetchFn: () => Promise<T>;
}): Promise<unknown> {
  const key = cacheKey(toolName, query);

  const [cached] = await db
    .select({ rawResponse: scrapeCache.rawResponse })
    .from(scrapeCache)
    .where(eq(scrapeCache.cacheKey, key))
    .limit(1);
  if (cached) return cached.rawResponse;

  const result = await fetchFn();
  const payload = result !== null && typeof result === "object" ? result : { raw: result };

  await db
    .insert(scrapeCache)
    .values({ cacheKey: key, toolName, rawResponse: payload })
    .onConflictDoNothing({ target: scrapeCache.cacheKey });

  return payload;
}

/** Models are asked for bare JSON but sometimes wrap it in ```json fences
 * anyway -- strip those before parsing rather than losing the whole response
 * to a parse error. */
export function parseJson<T>(text: string, fallback: T): T {
  const cleaned = text
    .trim()
    .replace(/^```(?:json)?\s*/i, "")
    .replace(/\s*```$/, "");
  try {
    return JSON.parse(cleaned) as T;
  } catch {
    return fallback;
  }
}

/** Cheap-model pass: raw scraped text -> structured JSON facts. */
export async function extractFacts(
  rawContext: string,
  extractionPrompt: string,
): Promise<unknown> {
  const content = await aimlClient.chat(
    [
      {
        role: "system",
        content: "Extract structured facts as compact JSON. Respond with JSON only.",
      },
      { role: "user", content: `${extractionPrompt}\n\n---\n${rawContext}` },
    ],
    { reasoning: false },
  );
  // The fast model wraps its JSON in ```json fences often enough that parsing
  // the bare string throws away most extractions -- the synthesis pass then
  // reasons over `{raw: "```json..."}` instead of real facts.
  return parseJson<unknown>(content, { raw: content });
}

/** Reasoning-model pass: structured facts + profile context -> final output. */
export function synthesize(systemPrompt: string, userPrompt: string): Promise<string> {
  return aimlClient.chat(
    [
      { role: "system", content: systemPrompt },
      { role: "user", content: userPrompt },
    ],
    { reasoning: true },
  );
}

File: backend/app/services/pipeline.py
"""Shared gather -> extract -> synthesize pipeline used by all four modules.

1. gather: call Bright Data (through cached_fetch, so repeat demo runs don't
   burn credits) to get raw web data.
2. extract: cheap AIML model turns raw data into structured facts.
3. synthesize: reasoning AIML model turns structured facts + profile context
   into the module's final output.
"""

import hashlib
import json
from collections.abc import Callable
from typing import Any

from sqlalchemy.orm import Session

from app.models import ScrapeCache
from app.services import aiml_client


def _cache_key(tool_name: str, query: str) -> str:
    return hashlib.sha256(f"{tool_name}:{query}".encode()).hexdigest()


def cached_fetch(db: Session, *, tool_name: str, query: str, fetch_fn: Callable[[], Any]) -> Any:
    """Look up (tool_name, query) in ScrapeCache; call fetch_fn and store on miss."""
    key = _cache_key(tool_name, query)
    cached = db.query(ScrapeCache).filter_by(cache_key=key).first()
    if cached:
        return cached.raw_response

    result = fetch_fn()
    payload = result if isinstance(result, (dict, list)) else {"raw": result}
    db.add(ScrapeCache(cache_key=key, tool_name=tool_name, raw_response=payload))
    db.commit()
    return payload


def extract_facts(raw_context: str, extraction_prompt: str) -> dict:
    """Cheap-model pass: raw scraped text -> structured JSON facts."""
    content = aiml_client.chat(
        [
            {"role": "system", "content": "Extract structured facts as compact JSON. Respond with JSON only."},
            {"role": "user", "content": f"{extraction_prompt}\n\n---\n{raw_context}"},
        ],
        reasoning=False,
    )
    try:
        return json.loads(content)
    except ValueError:
        return {"raw": content}


def synthesize(system_prompt: str, user_prompt: str) -> str:
    """Reasoning-model pass: structured facts + profile context -> final output."""
    return aiml_client.chat(
        [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ],
        reasoning=True,
    )
